### PR TITLE
Bug 1873043: Rebase to 1.19.2

### PR DESCRIPTION
**Pull request by tnozicka**
tracking sheet https://docs.google.com/spreadsheets/d/10KYptJkDB1z8_RYCQVBYDjdTlRfyoXILMa0Fg8tnNlY

TODO:
- [x] Update upstream tags on o/k repo (https://github.com/openshift/kubernetes/tree/v1.19.2)
- [x] Update `openshift/api` repository to v1.19.2 (https://github.com/openshift/api/pull/761)
- [x] Update `openshift/apiserver-library-go` to v1.19.2 (https://github.com/openshift/apiserver-library-go/pull/37)
- [x] Update `openshift/client-go` to v1.19.2 (https://github.com/openshift/client-go/pull/163)
- [x] Update `openshift/library-go` to v1.19.2 (https://github.com/openshift/library-go/pull/921)
- [x] Update the hyperkube image version to the release tag
- [x] Setup CI for AWS, GCP and Azure upgrades (https://github.com/openshift/release/pull/12618)
- [x] Setup self-upgrade test (the initial cluster is from this payload) (https://github.com/openshift/release/pull/12618)
- [x] Bump updated openshift dependencies
- [x] Update REBASE.openshift.md
- [x] Squash commits

Followups:
- [ ] Revendor to openshift/origin

/cc @marun @soltysh @sttts

### CODE REVIEW

**Comment by marun** [comment]
@tnozicka What does the TODO item 'create the branch' imply?

**Comment by tnozicka** [comment]
> @tnozicka What does the TODO item 'create the branch' imply?
We can either create the branch now and fast forward from master, or before the next rebase. In the general OCP flow we are pre-creating the branches but I don't feel strong either way.

/retest

**Comment by marun** [comment]
> We can either create the `origin-4.6-kubernetes-1.19.2` branch now and fast forward from master, or before the next rebase. In the general OCP flow we are pre-creating the branches but I don't feel strong either way.
> 
> Also I am not sure how useful the branch granularity is now with go mod, almost feels like all we need is `origin-4.6-kubernetes-1.19`. You can see a particular minor by tag and we are not using it for pinning anymore.

As per the rebase instructions I provided, rebases for 4.6 and beyond use regular PR flow.  Merge a rebase to o/k master or release branch, update vendoring in the related origin branch via SHA. Rebase branches are no longer required.

**Comment by tnozicka** [comment]
> As per the rebase instructions I provided, rebases for 4.6 and beyond use regular PR flow. Merge a rebase to o/k master or release branch, update vendoring in the related origin branch via SHA. Rebase-specific branches are no longer involved.

Sure, if we use release-4.6 (when master changes) we must disable the fast forward though. (Maybe you already did.) Rebase and release cut happen at different time. I haven't check yet, nor had time to look at the units because other stuff on fire today but I'll get to it.

**Comment by marun** [comment]
> Sure, if we use release-4.6 (when master changes) we must disable the fast forward though. (Maybe you already did.) Rebase and release cut happen at different time. I haven't check yet, nor had time to look at the units because other stuff on fire today but I'll get to it.

Why would it be necessary to disable ff? Regular PR rules apply: master will ff to release-4.6 until code freeze and when master re-opens it will be ff'ing to release-4.7. If you haven't landed 1.19.2 by the time of the cut, you'll switch the target branch of this PR from master to release-4.6.

**Comment by tnozicka** [comment]
> @tnozicka The SNI unit tests are flakey. As per slack conversations manually picking Jordan's unit test cleanup is suggested, though I don't recall seeing anything SNI-specific. It may be an openshift-specific flake.

it looks like a data race coming from our carry https://github.com/openshift/kubernetes/commit/538d9c4189431a192850ccf582219c17962e610b#diff-8c3b369a56b450db4f2c3715bec58edcR357

> I don't follow, why would it be necessary to disable ff?
I don't follow either, looks like I had a bad day seeing issues where there aren't any. Sorry about that.

File: build/dependencies.yaml [hunk 1]
```diff
@@ -74,7 +74,7 @@ dependencies:
 
   # etcd
   - name: "etcd"
-    version: 3.4.9
+    version: 3.4.13
```
Comment (tnozicka) [inline build/dependencies.yaml:77]: cc @hexfusion

**Comment by tnozicka** [comment]
@sttts merged 2 PRs to fix the race in the meantime so units are now passing without the additional carry to disable parallel run there

**Comment by marun** [comment]
@tnozicka Where are the carries? Also, a link to a spreadsheet indicating what you're picking and not picking. CI passing has limited value with the bare upstream tag since that's not anywhere close to what we'll merge.

**Comment by tnozicka** [comment]
needs deps merged, and then a real bump and squashing but otherwise is ready

File: OWNERS [hunk 2]
```diff
@@ -3,25 +3,22 @@
...
-    - sjenning
-    - smarterclayton
+    - sttts
     - soltysh
-    - tbielawa
+    - mfojtik
+    - marun
+    - tnozicka
+
+    # approvers are limited to to the team that manages rebases and pays the price for carries that are introduced
```
Comment (tnozicka) [inline OWNERS:13]: let's call this one out explicitly so it isn't overlooked. if something is on fire and none of us is reachable, architects can edit the labels

**Comment by marun** [comment]
@tnozicka Do you need to merge the lib bumps in order and revendor at every step, or have I been doing it wrong? I usually WIP or /hold subsequent lib bumps to ensure they aren't merged accidentally.

**Comment by tnozicka** [comment]
Some depend on each other, but just bumping kube in all of them shouldn't make them incompatible and I think that would be found out after a fake bump to o/k.

They also individually overwrite kube deps, so bumping in step probably doesn't make much of a difference except when you hit a conflict in the fake bump.

But I haven't done it previously to have a data point one way or the other. I didn't see anything about doing it one at the time in the instructions. Have you seen it fail previously and decided to do it in step or you were just being cautious?

**Comment by marun** [comment]
> But I haven't done it previously to have a data point one way or the other. I didn't see anything about doing it one at the time in the instructions. Have you seen it fail previously and decided to do it in step or you were just being cautious?

I've definitely seen cases where more changes were involved that required the PRs to merge in serial. If the bump is limited to changing accounting files (go.mod, go.sum, vendor.txt) then it's probably unnecessary, but anything outside of that I would be cautious.

That said, I recall @sttts indicating that shared lib bumps for minor rebases may not even be necessary.

**Comment by tnozicka** [comment]
sure, if you make incompatible changes in one, that would affect the other, then bumping serially make total sense. My point is that you'd find it out in this PR with the fake bump because the test would fail, right?

**Comment by marun** [comment]
> sure, if you make incompatible changes in one, that would affect the other, then bumping serially make total sense. My point is that you'd find it out in this PR with the fake bump because the test would fail, right?

Yes, the rebase PR is likely (but not guaranteed) to catch inconsistencies. Given the chain of dependency (client-go depends on api, library-go depends on api and client-go, apiserver-library-go depends on api, client-go and library-go), not merging the changes serially means there is a non-zero chance that one of the bumps will end up missing changes from one of its dependencies. The only way to guarantee consistency is to merge serially.

**Comment by tnozicka** [comment]
/hold
prow forgot to run selfupgrade and azure uppgrade test, engaging DPTP to trace why

**Comment by tnozicka** [comment]
/hold cancel
fixed the release config in https://github.com/openshift/release/pull/12975, the initial runs here were a prow bug

**Comment by tnozicka** [comment]
ci infra flake
/test e2e-azure-upgrade
/test e2e-cmd

**Comment by tnozicka** [comment]
extended upgrade tests are all green. ready to be reviewed before I squash it

**Comment by soltysh** [comment]
I have one question about: [UPSTREAM: <carry>: autoscaling: hack in working scaling for legacy deploymentconfig](https://github.com/openshift/kubernetes/commit/1145e0aefe9?w=1) which I left in the doc, the rest lgtm.

**Review by soltysh — APPROVED** [review]

**Comment by marun** [comment]
gcp jobs are failing due to quota issues, investigation ongoing.

**Comment by tnozicka** [comment]
another shot at gcp
/retest

**Comment by tnozicka** [comment]
broken still it seems
```
error: failed to load configuration: failed to make request to configresolver: Get "https://config.ci.openshift.org/config?branch=master&org=openshift&repo=kubernetes": dial tcp: lookup config.ci.openshift.org on 172.30.0.10:53: read udp 10.130.4.41:38326->172.30.0.10:53: i/o timeout
```

File: OWNERS (hunk 2 above)
Comment (sttts) [reply OWNERS:13]: :+1:

File: REBASE.openshift.md [hunk 3]
```diff
@@ -206,36 +231,24 @@ should be true:
...
-PR should not be merged until additional upgrade testing initiated with
-cluster-bot is passing:
-
-- [ ] `test upgrade [previous release e.g. 4.6] openshift/kubernetes#[PR#] [aws|azure]`
-       - Only gcp upgrades are tested automatically via presubmit
-         (`e2e-gcp-upgrade`) and it's necessary to manually test aws and azure.
-- [ ] `test upgrade openshift/kubernetes#[PR#] openshift/kubernetes#[PR#]`
-      - This 'self-upgrade' ensures that it is possible to upgrade _from_ the
-        rebased release. The other upgrade testing validates that it's possible
-        to upgrade _to_ the rebased release.
```
Comment (sttts) [inline REBASE.openshift.md:231]: why is all this gone? We do that automatically?

Comment (tnozicka) [reply REBASE.openshift.md:231]: yep, I've added all of this to CI and it is triggered by changing the kube version in the dockerfile and run automatically. you can see the jobs passing in status

Comment (tnozicka) [reply REBASE.openshift.md:231]: e.g. https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_kubernetes/361/pull-ci-openshift-kubernetes-master-e2e-aws-selfupgrade/1320659343942619136

File: openshift-kube-apiserver/authentication/oauth/tokenauthenticator.go [hunk 4]
```diff
@@ -2,8 +2,11 @@ package oauth
 
 import (
 	"context"
+	"crypto/sha256"
```
Comment (sttts) [inline openshift-kube-apiserver/authentication/oauth/tokenauthenticator.go]: squash with `UPSTREAM: <carry>: oauth-token-authn: switch to sha256~ prefix`

Comment (tnozicka) [reply openshift-kube-apiserver/authentication/oauth/tokenauthenticator.go]: yep, i think this is marked in the sheet to be squashed

**Comment by tnozicka** [comment]
Down to 66 carries from 112.
https://github.com/openshift/kubernetes/compare/86ec9cb59ffb2d19e804dfcc1ebb753c39d3ac0e..8d10f066bbceebbc367179de2206a126a6bcb5a3 proves it didn't change any code.

**Comment by tnozicka** [comment]
squashed the OWNERS changes so we are down to 65 :)
again noop https://github.com/openshift/kubernetes/compare/8d10f066bbceebbc367179de2206a126a6bcb5a3..9a570ac82c8b40743032295452b437c11cf3f364

**Comment by sttts** [comment]
Lgtm
/approve

Leaving final lgtm to @soltysh

**Comment by tnozicka** [comment]
Fixed our OWNERS file to apply globally https://github.com/openshift/kubernetes/compare/9a570ac82c8b40743032295452b437c11cf3f364..fa45d8a287b15c800064f4274cb38c70bcace6c0

**Comment by tnozicka** [comment]
@soltysh per your request I've squashed `UPSTREAM: <carry>: Supply context to API calls` into the individual commits
```
export patch_files=$( git diff-tree --no-commit-id --name-only --diff-filter=d -r a0e976788  | tr "\n" " " ) && git rebase 4d517425f012822b4452a2da21b43fef04152ae4 -i -x 'set -eEuxo pipefail; files=$( git diff-tree --no-commit-id --name-only --diff-filter=d -r HEAD | tr "\n" " " ) && for f in ${files}; do if [[ -n $( echo ${patch_files} | grep "${f}" ) ]]; then git show a0e976788 -- "${f}" | ( git apply && git add "${f}" && git commit --amend --no-edit ) || true; fi; done'
```
noop as per https://github.com/openshift/kubernetes/compare/fa45d8a287b15c800064f4274cb38c70bcace6c0..251732001dca9f398fd69a8c36fd3c1603b83317

**Comment by tnozicka** [comment]
fixed a typo https://github.com/openshift/kubernetes/compare/80e50d20c685504237bdc055f456aec5a1f5311c..512f733fcbf275af939217b1a566b831775ba86e otherwise its ready

**Pull request merged**